### PR TITLE
Fix T5 incorrect weight decay in Trainer and official summarization example

### PR DESCRIPTION
**Pull request by ADAning**
# What does this PR do?

Official summrization examples use T5 as pretrained model, but the name of T5 layer norm is `layer_norm`, not `layerNorm`:
```
from transformers import T5Tokenizer, T5ForConditionalGeneration
model = T5ForConditionalGeneration.from_pretrained("t5-small")
for n, p in model.named_parameters():
    print(n)
```
Output:
```
...
encoder.block.0.layer.0.layer_norm.weight
encoder.block.0.layer.1.DenseReluDense.wi.weight
encoder.block.0.layer.1.DenseReluDense.wo.weight
encoder.block.0.layer.1.layer_norm.weight
...
```
In [Official example of summarization](https://github.com/huggingface/transformers/blob/49c8c67fb815a277405f84dea4a66353e19fb347/examples/pytorch/summarization/run_summarization_no_trainer.py#L529), `layer_norm` not included:
```
no_decay = ["bias", "LayerNorm.weight"]
```

A similar problem occurred in [trainer.py](https://github.com/huggingface/transformers/blob/49c8c67fb815a277405f84dea4a66353e19fb347/src/transformers/trainer.py#L970), which may cause `Seq2SeqTrainer` train T5 layer norm with weight decay:
```
if self.optimizer is None:
            decay_parameters = get_parameter_names(opt_model, [nn.LayerNorm])
            decay_parameters = [name for name in decay_parameters if "bias" not in name]
```
Because T5 uses [T5LayerNorm](https://github.com/huggingface/transformers/blob/49c8c67fb815a277405f84dea4a66353e19fb347/src/transformers/models/t5/modeling_t5.py#L239) Layer norm not`nn.LayerNorm`:

```
class T5LayerNorm(nn.Module):
    def __init__(self, hidden_size, eps=1e-6):
        """
        Construct a layernorm module in the T5 style. No bias and no subtraction of mean.
        """
        super().__init__()
        self.weight = nn.Parameter(torch.ones(hidden_size))
        self.variance_epsilon = eps
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: examples/pytorch/summarization/run_summarization_no_trainer.py [hunk 1]
```diff
@@ -526,7 +526,7 @@ def postprocess_text(preds, labels):
 
     # Optimizer
     # Split weights in two groups, one with weight decay and the other not.
-    no_decay = ["bias", "LayerNorm.weight"]
+    no_decay = ["bias", "layer_norm.weight"]
```
Comment (sgugger) [inline examples/pytorch/summarization/run_summarization_no_trainer.py:529]: We should keep both here, since the user might use other models.

**Review by sgugger — COMMENTED** [review]
Thanks for your PR, I've added a few comments.

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -70,6 +70,7 @@
 from .modelcard import TrainingSummary
 from .modeling_utils import PreTrainedModel, load_sharded_checkpoint, unwrap_model
 from .models.auto.modeling_auto import MODEL_FOR_CAUSAL_LM_MAPPING_NAMES
+from .models.t5.modeling_t5 import T5LayerNorm
```
[the remark this answers is not in the document]
Comment (ADAning) [reply src/transformers/trainer.py]: Thanks for your comments! It seems that no one has written any types of layer norm as subclass  of `nn.LayerNorm`. I'm not sure if this is true:
```python
class T5LayerNorm(nn.LayerNorm):  
    def __init__(self, hidden_size, eps):  
        """  
        Construct a layernorm module in the T5 style. No bias and no subtraction of mean.        """        super().__init__(hidden_size, eps, elementwise_affine=False)  
        self.weight = nn.Parameter(torch.ones(hidden_size))  
        self.variance_epsilon = eps  
  
        self.reset_parameters()  
  
    def forward(self, hidden_states):  
  
        # T5 uses a layer_norm which only scales and doesn't shift, which is also known as Root Mean  
        # Square Layer Normalization https://arxiv.org/abs/1910.07467 thus varience is calculated        # w/o mean and there is no bias. Additionally we want to make sure that the accumulation for        # half-precision inputs is done in fp32  
        variance = hidden_states.to(torch.float32).pow(2).mean(-1, keepdim=True)  
        hidden_states = hidden_states * torch.rsqrt(variance + self.variance_epsilon)  
  
        # convert into half-precision if necessary  
        if self.weight.dtype in [torch.float16, torch.bfloat16]:  
            hidden_states = hidden_states.to(self.weight.dtype)  
  
        return self.weight * hidden_states  
  
    def reset_parameters(self):  
        if self.weight is not None:  
            nn.init.ones_(self.weight)
```
`LongT5LayerNorm` is copied from `T5LayerNorm` which need to be changed same.

File: src/transformers/models/longt5/modeling_longt5.py [hunk 3]
```diff
@@ -221,15 +221,17 @@ def _create_global_aggregates(
 
 
 # Copied from transformers.models.t5.modeling_t5.T5LayerNorm with T5->LongT5
-class LongT5LayerNorm(nn.Module):
+class LongT5LayerNorm(nn.LayerNorm):
```
Comment (patrickvonplaten) [inline src/transformers/models/longt5/modeling_longt5.py:224]: This would create unused bias weights for every layernorm in T5 at the moment, I don't think we want this. @sgugger I'm actually not really in favor of adding this abstraction here to the `T5LayerNorm` class

Comment (sgugger) [reply src/transformers/models/longt5/modeling_longt5.py:224]: Only if the init calls the superclass. If it calls the super of the super class (nn.Module), there is nothing breaking. We just can't have the Trainer start depending on multiple modeling files because there are 100 flavors of LayerNorms.

An alternative would be to have a constant in `pt_utils` where each submodule that defines a specific LayerNorm adds their own, would that work better?

So here we would have after defining `LongT5LayerNorm`:
```
ALL_LAYERNORM_LAYERS.append(LongT5LayerNorm)
```

Comment (sgugger) [reply src/transformers/models/longt5/modeling_longt5.py:224]: @ADAning Could you amend your PR to use this solution maybe?

Comment (ADAning) [reply src/transformers/models/longt5/modeling_longt5.py:224]: Does `pt_utils` mean `transformers.utils` ?

> @ADAning Could you amend your PR to use this solution maybe?

File: src/transformers/models/t5/modeling_t5.py [hunk 4]
```diff
@@ -262,6 +262,8 @@ def forward(self, hidden_states):
         return self.weight * hidden_states
 
 
+ALL_LAYERNORM_LAYERS.append(T5LayerNorm)
```
Comment (sgugger) [inline src/transformers/models/t5/modeling_t5.py:265]: This should be after the try except block below, which may change `T5LayerNorm`.

File: src/transformers/models/longt5/modeling_longt5.py [hunk 5]
```diff
@@ -247,6 +248,8 @@ def forward(self, hidden_states):
         return self.weight * hidden_states
 
 
+ALL_LAYERNORM_LAYERS.append(LongT5LayerNorm)
```
Comment (sgugger) [inline src/transformers/models/longt5/modeling_longt5.py:251]: This should be after the try/except block below, which can change `LongT5LayerNorm`.

**Review by sgugger — APPROVED** [review]
Just re-opened your PR! In general you only need to rebase if the PR has gone a long time (a month) or if there is some critical changes in the main branch you need, so don't bother for small changes :-)

**Comment by sgugger** [comment]
Thanks a lot for iterating with us!

**Review by patrickvonplaten — COMMENTED** [review]
Thanks a lot for iterating on this and sorry to reply so late here - the final solution looks very nice!
